src/database/mod.rs

```diff
@@ -23,11 +32,55 @@ use crate::DbErr;
 #[derive(Debug, Default)]
 pub struct Database;
 
+/// Supported database kinds of [sqlx::ConnectOptions]'.
+#[derive(Debug, Clone)]
+pub enum SqlxConnectOptions {
+    #[cfg(feature = "sqlx-mysql")]
+    MySql(MySqlConnectOptions),
+    #[cfg(feature = "sqlx-postgres")]
+    Postgres(PgConnectOptions),
+    #[cfg(feature = "sqlx-sqlite")]
+    Sqlite(SqliteConnectOptions),
+    #[cfg(feature = "mock")]
+    Mock(DbBackend),
+}
+
+impl SqlxConnectOptions {
+    /// The database backend type
+    pub fn get_db_backend_type(&self) -> DbBackend {
+        match self {
+            #[cfg(feature = "sqlx-mysql")]
+            SqlxConnectOptions::MySql(_) => DbBackend::MySql,
+            #[cfg(feature = "sqlx-postgres")]
+            SqlxConnectOptions::Postgres(_) => DbBackend::Postgres,
+            #[cfg(feature = "sqlx-sqlite")]
+            SqlxConnectOptions::Sqlite(_) => DbBackend::Sqlite,
+            #[cfg(feature = "mock")]
+            SqlxConnectOptions::Mock(db_backend) => db_backend.clone(),
+        }
+    }
+
+    #[cfg(feature = "mock")]
+    pub fn mock(db_backend: DbBackend) -> SqlxConnectOptions {
+        Self::Mock(db_backend)
+    }
+
+    #[cfg(feature = "mock")]
+    pub fn is_mock(&self) -> bool {
+        match self {
+            SqlxConnectOptions::Mock(_) => true,
+            _ => false,
+        }
+    }
+}
+
 /// Defines the configuration options of a database
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct ConnectOptions {
-    /// The URI of the database
-    pub(crate) url: String,
+    /// The database sqlx::ConnectOptions used to connect to the database.
+    pub(crate) connect_options: SqlxConnectOptions,
+    /// The URI of the database, if this struct was created from an URI string, otherwise None
+    pub(crate) url: Option<String>,
```

Comment:
We should make this an enum with two variants
- `SqlxConnectOptions`: wrapper enum of all supported databases' connection option
- `ConnectionOptions`: raw url with some other connection option included in the old `ConnectOptions` struct

Reply:
> This PR has changed the url() in a way that it will only return something if it was created from an url because otherwise we currently cannot get the url from the connect_options.

Which is pretty confusing IMO... :\

Reply:
What do you think we should do with the url? Should we keep the url() method? Should that only return the url if the sea_orm::ConenctOptions was created from an url string?

Reply:
I think we can keep the `ConnectOptions` untouched. And introduce a new trait (e.g. `ConnectionOptionTrait`) that convert any kinds of DatabaseOptions into `Database`. Then, `Database::connect()` method takes any `ConnectionOptionTrait`.

Make sense? CC @tyt2y3

Reply:
@billy1624 No this doesn't answer the question about what should we return from the `DatabaseConnection` when get_url is called if the `DatabaseConnection` was created from a `sqlx::ConnectOptions` implementation like `SqliteConnectOptions?

About `ConnectionOptionTrait` we don't need that to be able to implement creating DatbaseConnections from both urls and both sqlx::ConnectOptions implementations. You can just do what I already did in this PR implementing with `TryFrom`.

Reply:
@billy1624 any update about this? Specifically about what should `get_url` return if the `DatabaseConnection` was not created from an url string but from an `sqlx::ConnectOptions` e.g.: `SqliteConnectOptions`?

Reply:
@billy1624 @tyt2y3 Anyone?

Reply:
Hey @05storm26, sorry for the delay. I will get back to you ASAP

Reply:
Hi @billy1624 the required sqlx commits have been merged so this could be merged after the review is completed. 

Could you take a look at this conversation again?